apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -295,590 +412,954 @@ poll(Node, ReturnAddr, Shard, Iterator, Opts = #{timeout := Timeout}) ->
+            %% variables:
+            FlowControl = atomics:new(3, []),
+            SubState = #sub_state{
+                req_id = SubRef,
+                client = Client,
+                mref = MRef,
+                flowcontrol = {MaxUnacked, FlowControl},
+                rank = Rank,
+                stream = Stream,
+                topic_filter = TF,
+                start_key = DSKey,
+                it = It,
+                msg_matcher = MsgMatcher
+            },
+            %% Store another reference to the flow control object in a
+            %% different table readable by the client during suback:
+            ets:insert(fc_tab(Shard), #fctab{
+                sub_ref = SubRef, max_unacked = MaxUnacked, aref = FlowControl
+            }),
+            %% Create an entry in the owner table:
+            ets:insert(owner_tab(Shard), #owner_tab{sub_ref = SubRef, mref = MRef}),
+            %% Schedule dispaching to the catchup worker:
+            D = D0#d{pending = [SubState | Pending]},
+            Reply = {reply, From, {ok, SubRef}},
+            case State of
+                ?idle ->
+                    {next_state, ?busy, D, Reply};
+                _ ->
+                    {keep_state, D, Reply}
+            end;
+        Err = {error, _, _} ->
+            {keep_state_and_data, {reply, From, Err}}
+    catch
+        EC:Reason:Stack ->
+            Error = ?err_unrec({EC, Reason, Stack}),
+            {keep_state_and_data, {reply, From, Error}}
+    end;
+%% Handle unsubscribe call:
+handle_event(
+    {call, From},
+    #unsub_req{id = SubId},
+    _State,
+    D0
+) ->
+    {Ret, D} = remove_subscription(SubId, D0),
+    {keep_state, D, {reply, From, Ret}};
+%% Handle disown request:
+handle_event(
+    cast,
+    #disown_stuck_req{sub_state = SubState},
+    _State,
+    #d{dbshard = DBShard}
+) ->
+    #sub_state{req_id = SubRef} = SubState,
+    case ets:update_element(owner_tab(DBShard), SubRef, {#owner_tab.owner, undefined}) of
         true ->
-            emqx_ds_builtin_metrics:inc_poll_requests_dropped(Metrics, 1),
-            Reply = {error, recoverable, too_many_requests},
-            {reply, Reply, S};
+            %% Subscription was found, insert it to the disowned
+            %% table:
+            true = ets:insert_new(disowned_sub_tab(DBShard), SubState),
+            ok;
         false ->
-            ets:insert(S#s.pending_queue, Req),
-            {reply, ok, ensure_fulfill_loop(S)}
+            %% Subscription has been removed in the meantime. Ignore.
+            ok
+    end,
+    keep_state_and_data;
+%% Handle wakeup subscription call:
+handle_event(
+    {call, From},
+    #wakeup_sub_req{id = SubId},
+    State,
+    D0 = #d{pending = Pending, stuck_sub_tab = Tab}
+) ->
+    case ets:take(Tab, SubId) of
+        [SubS = #sub_state{flowcontrol = {_MaxUnacked, ARef}}] ->
+            atomics:put(ARef, ?fc_idx_stuck, 0),
```

Comment:
It seems this could race with the `atomics:put(_, stuck, 1)` from `keep_and_seqno` which is called externally via `beams_conclude/3`?

Reply:
What is the precise scenario for the race condition? 

This event is _only_ triggered by a client that _knows_ that its subscription is stuck (i.e. ?fc_idx_stuck` = 1).